rust_src/src/indent.rs

```diff
@@ -179,4 +180,127 @@ pub fn indent_to(column: EmacsInt, minimum: Option<EmacsInt>) -> EmacsInt {
     mincol
 }
 
+/// Scan through the current buffer, calculating screen position.
+/// Scan the current buffer forward from offset FROM, assuming it is
+/// at position FROMPOS--a cons of the form (HPOS . VPOS)-- to
+/// position TO or position TOPOS--another cons of the form (HPOS
+/// . VPOS)-- and return the ending buffer position and screen
+/// location.
+///
+/// If TOPOS is nil, the actual width and height of the window's text
+/// area are used.
+///
+/// There are three additional arguments:
+///
+/// WIDTH is the number of columns available to display text; this
+/// affects handling of continuation lines.  A value of nil
+/// corresponds to the actual number of available text columns.
+///
+/// OFFSETS is either nil or a cons cell (HSCROLL . TAB-OFFSET).
+/// HSCROLL is the number of columns not being displayed at the left
+/// margin; this is usually taken from a window's hscroll member.
+/// TAB-OFFSET is the number of columns of the first tab that aren't
+/// being displayed, perhaps because the line was continued within it.
+/// If OFFSETS is nil, HSCROLL and TAB-OFFSET are assumed to be zero.
+///
+/// WINDOW is the window to operate on.  It is used to choose the
+/// display table; if it is showing the current buffer, it is used
+/// also for deciding which overlay properties apply.  Note that
+/// `compute-motion' always operates on the current buffer.
+///
+/// The value is a list of five elements:
+///   (POS HPOS VPOS PREVHPOS CONTIN)
+/// POS is the buffer position where the scan stopped.
+/// VPOS is the vertical position where the scan stopped.
+/// HPOS is the horizontal position where the scan stopped.
+///
+/// PREVHPOS is the horizontal position one character back from POS.
+/// CONTIN is t if a line was continued after (or within) the previous character.
+///
+/// For example, to find the buffer position of column COL of line
+/// LINE of a certain window, pass the window's starting location as
+/// FROM and the window's upper-left coordinates as FROMPOS.  Pass the
+/// buffer's (point-max) as TO, to limit the scan to the end of the
+/// visible section of the buffer, and pass LINE and COL as TOPOS.
+#[lisp_fn]
+pub fn compute_motion(
+    from: LispNumber,
+    frompos: LispCons,
+    to: LispNumber,
+    topos: LispObject,
+    width: Option<EmacsInt>,
+    offsets: LispObject,
+    window: LispWindowLiveOrSelected,
+) -> LispObject {
+    let from_hpos = frompos.car().into();
+    let from_vpos = frompos.cdr().into();
+
+    let mut win: LispWindowRef = window.into();
+    let window_width = EmacsInt::from(win.body_width(false))
+        - if cfg!(feature = "window-system") && win.get_frame().is_gui_window() {
+            0
+        } else {
+            1
+        };
+
+    let (to_hpos, to_vpos) = if topos.is_nil() {
+        (
+            window_width,
+            EmacsInt::from(unsafe { window_internal_height(win.as_mut()) }),
```

Comment:
`window_internal_height` should be a member of `LispWindowRef`, something like `internal_height()`. The needed C function can be a light wrapper around it. here we already have a LispWindowRef so we can use the member directly and get rid of both the unsafe and the EmacsInt conversion.

Reply:
I ported it. How is the integer conversion skippable though? All window size related functions use `i32`.